File: problem11/typescript/matrixProduct.test.ts
import { describe, it, expect, beforeAll } from 'vitest';

let greatestProduct: typeof import('./matrixProduct').greatestProduct;
type Coord = import('./matrixProduct').Coord;
type Matrix = import('./matrixProduct').Matrix;

const N = 4;
const ADJ = 4;

const matrix: Matrix = [
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12],
    [13, 14, 15, 16]
];

beforeAll(async () => {
    const mod = await import('./matrixProduct');
    greatestProduct = mod.greatestProduct;
});

describe('greatestProduct', () => {
    it('finds the greatest product in a 4x4 matrix', () => {
        const coords: Coord[] = [];
        const result = greatestProduct(matrix, coords, 4);
        expect(result).toBe(43680); // 13*14*15*16
        expect(coords).toEqual([
            [3, 0], [3, 1], [3, 2], [3, 3]
        ]);
    });
});

File: problem11/go/matrix_product_test.go
package main

import (
	"testing"
	"reflect"
)

func TestGreatestProduct(t *testing.T) {
	matrix := [][]int{
		{1, 2, 3, 4},
		{5, 6, 7, 8},
		{9, 10, 11, 12},
		{13, 14, 15, 16},
	}
	coords := make([][2]int, 0)
	result := GreatestProduct(matrix, &coords, 4)

	expected := 43680 // 13*14*15*16
	expectedCoords := [][2]int{{3, 0}, {3, 1}, {3, 2}, {3, 3}}

	if result != expected {
		t.Errorf("Expected product %d, got %d", expected, result)
	}
	if !reflect.DeepEqual(coords, expectedCoords) {
		t.Errorf("Expected coords %v, got %v", expectedCoords, coords)
	}
}
